### PR TITLE
Geometry checking and including spherical models

### PR DESCRIPTION
**Pull request by bobmyhill**
This is a continuation of PR #1685. Additionally, the included GeometryModels in several files were inconsistent; this PR adds Sphere to boundary_composition/spherical_constant.cc and EllipsoidalChunk to initial_temperature/spherical_shell.cc. Tests are created for both of these new cases.

### CODE REVIEW

File: include/aspect/boundary_composition/spherical_constant.h [hunk 1]
```diff
@@ -87,6 +89,7 @@ namespace aspect
         /**
          * Compositions at the inner and outer boundaries.
          */
+        double surface_composition;
         double inner_composition;
```
Comment (bangerth) [inline include/aspect/boundary_composition/spherical_constant.h]: what does the new variable do? the class already has inner/outer boundary values, and the comment doesn't help either :-)

File: source/boundary_composition/spherical_constant.cc [hunk 2]
```diff
@@ -41,30 +42,42 @@ namespace aspect
...
       const GeometryModel::Interface<dim> *geometry_model = &this->get_geometry_model();
       const std::string boundary_name = geometry_model->translate_id_to_symbol_name(boundary_indicator);
 
-      if (dynamic_cast<const GeometryModel::SphericalShell<dim>*>(geometry_model) != 0
-          || dynamic_cast<const GeometryModel::Chunk<dim>*>(geometry_model) != 0
-          || dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*>(geometry_model) != 0)
+      if (dynamic_cast<const GeometryModel::Sphere<dim>*>(geometry_model) != 0)
+        {
+          if (boundary_name == "surface")
+            return surface_composition;
```
Comment (bangerth) [inline source/boundary_composition/spherical_constant.cc]: I think I would prefer if this returned "outer_composition", and that we document this in the definition of the input parameters

File: source/boundary_composition/spherical_constant.cc [hunk 3]
```diff
@@ -99,6 +120,9 @@ namespace aspect
       {
         prm.enter_subsection("Spherical constant");
         {
+          prm.declare_entry ("Surface composition", "0",
+                             Patterns::Double (),
+                             "Composition at the surface (lithosphere water/air). Units: none.");
```
Comment (bangerth) [inline source/boundary_composition/spherical_constant.cc]: my proposal would remove this parameter, and add a sentence to the definition of "Outer composition" to say that that is used for the surface when the geometry is a sphere.

File: source/initial_temperature/spherical_shell.cc [hunk 4]
```diff
@@ -38,42 +39,55 @@ namespace aspect
...
+          // TODO
+          // If the eccentricity of the EllipsoidalChunk is non-zero, the radius can vary along a boundary,
+          // but the maximal depth is the same everywhere and we could calculate a representative pressure
+          // profile. However, it requires some extra logic with ellipsoidal
+          // coordinates, so for now we only allow eccentricity zero.
+          // Using the EllipsoidalChunk with eccentricity zero can still be useful,
+          // because the domain can be non-coordinate parallel.
+
+          AssertThrow(gm->get_eccentricity() == 0.0,
+                      ExcMessage("This plugin cannot be used with a non-zero eccentricity. "));
```
Comment (bangerth) [inline source/initial_temperature/spherical_shell.cc]: This sounds more like you want to use `ExcNotImplemented()` given the TODO above

File: source/initial_temperature/spherical_shell.cc [hunk 5]
```diff
@@ -38,42 +39,55 @@ namespace aspect
...
+
+          R1 = gm->get_semi_major_axis_a();
+        }
+      else
+        {
+          Assert (false, ExcMessage ("This initial condition can only be used if the geometry "
+                                     "is a sphere, a spherical shell or a chunk."));
```
Comment (bangerth) [inline source/initial_temperature/spherical_shell.cc]: ...or ellipsoidal chunk

File: source/initial_temperature/spherical_shell.cc [hunk 6]
```diff
@@ -175,23 +188,64 @@ namespace aspect
...
+          // TODO
+          // If the eccentricity of the EllipsoidalChunk is non-zero, the radius can vary along a boundary,
+          // but the maximal depth is the same everywhere and we could calculate a representative pressure
+          // profile. However, it requires some extra logic with ellipsoidal
+          // coordinates, so for now we only allow eccentricity zero.
+          // Using the EllipsoidalChunk with eccentricity zero can still be useful,
+          // because the domain can be non-coordinate parallel.
+
+          AssertThrow(gm->get_eccentricity() == 0.0,
+                      ExcMessage("This plugin cannot be used with a non-zero eccentricity. "));
```
Comment (bangerth) [inline source/initial_temperature/spherical_shell.cc]: same here

File: source/initial_temperature/spherical_shell.cc [hunk 7]
```diff
@@ -175,23 +188,64 @@ namespace aspect
...
+          R0 = gm->get_semi_major_axis_a() - gm->maximal_depth();
+          R1 = gm->get_semi_major_axis_a();
+        }
+      else
+        {
+          Assert (false, ExcMessage ("This initial condition can only be used if the geometry "
+                                     "is a sphere, a spherical shell or a chunk."));
+
+          R0 = std::numeric_limits<double>::quiet_NaN();
+          R1 = std::numeric_limits<double>::quiet_NaN();
```
Comment (bangerth) [inline source/initial_temperature/spherical_shell.cc]: That's one way. You shouldn't get here, so it doesn't matter, but we've started using `numbers::signaling_nan<double>()` in such cases -- that *really* makes sure that nobody uses these values!

File: tests/time_dependent_temperature_bc.cc [hunk 8]
```diff
@@ -110,13 +110,13 @@ namespace aspect
...
     {
-      // verify that the geometry is in fact a time_dep_box since only
+      // verify that the geometry is a time_dep_box since only
       // for this geometry do we know for sure what boundary indicators it
       // uses and what they mean
       Assert (dynamic_cast<const GeometryModel::Box<dim>*>(&this->get_geometry_model())
               != 0,
               ExcMessage ("This boundary model is only implemented if the geometry is "
-                          "in fact a time_dep_box."));
+                          "a time_dep_box."));
```
Comment (bangerth) [inline tests/time_dependent_temperature_bc.cc]: You really didn't like the "in fact" in all of these places, did you? ;-)

Comment (bobmyhill) [inline tests/time_dependent_temperature_bc.cc]: :D

Comment (bobmyhill) [inline tests/time_dependent_temperature_bc.cc]: It's a personal bugbear, in fact!

**Comment by bangerth** [comment]
One of the tests also fails. Do you know why?

**Comment by bobmyhill** [comment]
Yes, this will be because I didn't have the test output until amending the "updated tests" commit.

**Comment by bangerth** [comment]
OK to merge.

**Pull request merged**